### PR TITLE
[WIP] EBNF of pact

### PR DESCRIPTION
**Pull request by jmcardon**

### CODE REVIEW

File: docs/Grammar.md [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+(see: https://www.haskell.org/happy/doc/html/sec-sequences.html).
+
+```ebnf
+(* A program is a collection of modules and expressions *)
+program        ::= top_level_list
+top_level_list ::=    module top_level_list | 
+                        expr top_level_list | 
+                   interface top_level_list | ε
+
+module ::= '(' 'module' ident kset doc_or_meta decl_list ')'
```
Comment (emilypi) [inline docs/Grammar.md]: For an example of a complex module with an interface, see here: 

https://github.com/kadena-io/chainweb-node/blob/master/pact/coin-contract/coin.pact
https://github.com/kadena-io/chainweb-node/blob/master/pact/coin-contract/coin-sig.pact

One thing that we'll need to spec are `defpact`s, which are multistep coroutines - I'll commit the code tomorrow modeling their use, but there's an example above in `coin.pact`.

File: docs/Grammar.md [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+pred                 ::= 'keys-all' | 'keys-1' | 'keys-2'
+
+numbers ::= number { number }
+letter ::= { (uc_letter | lc_letter | '_' ) }
+uc_letter ::= 'A' | ... | 'Z'
+lc_letter ::= 'a' | ... | 'z'
+number    ::= '0' | ... | '9'
+op ::= (* Printable operators *)
+string ::= (* All printable characters *)
+hash ::= (* Todo: is your hash hex encoded? *)
```
Comment (emilypi) [inline docs/Grammar.md]: base64-encoded

File: docs/Grammar.md [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+kset                 ::= hashes | '{' 'keys' ':' hashes '}' | {' 'keys' ':' hashes ',' 'pred' ':' pred '}'
+pred                 ::= 'keys-all' | 'keys-1' | 'keys-2'
+
+numbers ::= number { number }
+letter ::= { (uc_letter | lc_letter | '_' ) }
+uc_letter ::= 'A' | ... | 'Z'
+lc_letter ::= 'a' | ... | 'z'
+number    ::= '0' | ... | '9'
+op ::= (* Printable operators *)
+string ::= (* All printable characters *)
```
Comment (emilypi) [inline docs/Grammar.md:94]: There's no spec for string chars and op chars, and we should decide on one. I'm leaning towards unicode being  the standard so we can support multiple languages

Comment (jmcardon) [reply docs/Grammar.md:94]: :+1: sounds good. I honestly left it out since I wasn';t sure what you wanted to do here.

File: docs/Grammar.md [hunk 4]
```diff
@@ -23,6 +23,8 @@ decls ::=  '(' decl ')' { '(' decl ')' }
 decl      ::= defun | defschema | defconst | defcap | defpact |
               deftable | bless
 
+module_guard ::= ident | kset
```
Comment (jmcardon) [inline docs/Grammar.md:26]: ah, so you can define it inline or point it to an existing keyset, correct?

**Pull request merged**